File: c/src/pe1.c
#define _CRT_SECURE_NO_WARNINGS 1

#include <pe1.h>

#include <inttypes.h>
#include <stdio.h>

#define SUM_X2(n) ((n) * ((n) + 1))

void pe1(uint32_t n) {
  uint32_t sum = 0;
  if (n > 2) {
    sum = (3 * SUM_X2(n / 3) + 5 * SUM_X2(n / 5) - 15 * SUM_X2(n / 15)) >> 1;
  }

  printf("%" PRIu32 " below %" PRIu32 "\n", sum, n);
}

void pe1_loop(uint32_t n) {
  uint32_t sum = 0;
  if (n > 2) {
    for (uint32_t i = 3; i < n + 1; ++i) {
      if (i % 3 == 0 || i % 5 == 0) {
        sum += i;
      }
    }
  }

  printf("%" PRIu32 " below %" PRIu32 "\n", sum, n);
}

int pe1_main(void) {
  uint32_t n;

  while (1) {
    printf("below: ");
    if (scanf("%" PRIu32, &n) != 1) {
      scanf("%*s");
      puts("Input Number.");
    } else {
      if (!n) {
        break;
      }
      pe1(n);
    }
  }

  return 0;
}

File: go/pe1.go
package main

import (
    "fmt"
    "os"
    "strconv"
)

func sumx2(n uint64) uint64 {
    return n * (n + 1)
}

func pe1(n uint64) uint64 {
    return (3 * sumx2(n / 3) + 5 * sumx2(n / 5) - 15 * sumx2(n / 15)) / 2
}

func pe1loop(n uint64) uint64 {
    var sum uint64 = 0
    for i := uint64(1); i <= n; i++ {
        if i % 3 == 0 || i % 5 == 0 {
            sum += i
        }
    }
    
    return sum
}

func main() {
    var n uint64
    if len(os.Args) > 1 {
        var err error
        n, err = strconv.ParseUint(os.Args[1], 10, 64)
        if err != nil {
            fmt.Println("failed ParseUint: ", err)
        }
    } else {
        n = 1000
    }
    
    ans := pe1(n)
    fmt.Println("pe1: ", ans)
}
